Document InaccessibleObjectException/IllegalAccessException on JDK 16+ with solution

sergey-volkov-lm opened the pull request:
Some users face InaccessibleObjectException/IllegalAccessException working on JDK 16+.  
As far as I'm concerned, this is not considered as a bug in mockk; or fix is just not planned yet.  
So I added separate MD file describing the problem and link to it from main README.md (section "known issues")  

I'm not sure where to put this MD file, I supposed that `doc/md` directory is fine.

### Linked issues
* https://github.com/mockk/mockk/issues/681
* https://github.com/mockk/mockk/issues/929
* https://github.com/mockk/mockk/issues/864
* https://github.com/mockk/mockk/issues/834
* https://github.com/Ninja-Squad/springmockk/issues/65

aSemy commented on doc/md/jdk16-access-exceptions.md:
````diff
@@ -0,0 +1,62 @@
+}
+```
+
+For maven users:
````
```suggestion
For Maven users:
```

aSemy commented on doc/md/jdk16-access-exceptions.md:
```diff
@@ -0,0 +1,62 @@
+Add JVM argument `--add-opens java.base/java.time=ALL-UNNAMED` _for each package_ you need  
+Description: https://docs.oracle.com/en/java/javase/16/migrate/migrating-jdk-8-later-jdk-releases.html
+
+For gradle users:
```
```suggestion
For Gradle users:
```

aSemy commented on doc/md/jdk16-access-exceptions.md:
```diff
@@ -0,0 +1,62 @@
+
+### Problem symptoms 
+On JDK 16 and above you may encounter InaccessibleObjectException or IllegalAccessException in following known cases:
+1. _Some_ usages of `mockkStatic` on java standard library classes, e.g. 
```
```suggestion
1. _Some_ usages of `mockkStatic` on Java standard library classes, e.g. 
```

aSemy commented on doc/md/jdk16-access-exceptions.md:
````diff
@@ -0,0 +1,62 @@
+     module java.base does not "opens java.time" to unnamed module @4501b7af
+   ```
+
+2. Spying on java standard library classes, e.g.
````
```suggestion
2. Spying on Java standard library classes, e.g.
```

aSemy commented on doc/md/jdk16-access-exceptions.md:
```diff
@@ -0,0 +1,62 @@
+# JDK 16+ access exceptions
+
+### Problem symptoms 
+On JDK 16 and above you may encounter InaccessibleObjectException or IllegalAccessException in following known cases:
```
```suggestion
On JDK 16 and above you may encounter `InaccessibleObjectException` or `IllegalAccessException` in following known cases:
```

aSemy commented on README.md:
```diff
@@ -33,6 +33,7 @@ From version 1.13.0 MockK supports Kotlin 1.4 and higher
  
 * PowerMock needs a workaround to run together with MockK [#79](https://github.com/mockk/mockk/issues/79#issuecomment-437646333). (not sure after workaround if it is generally usable or not, please somebody report it)
 * Inline functions cannot be mocked: see the discussion on [this issue](https://github.com/mockk/mockk/issues/27)
+* Spies, mockkStatic may not work on JDK 16+; `InaccessibleObjectException`/`IllegalAccessException`: [read more here](doc/md/jdk16-access-exceptions.md)
```
```suggestion
* Spies, `mockkStatic` may not work on JDK 16+; `InaccessibleObjectException`/`IllegalAccessException`: [read more here](doc/md/jdk16-access-exceptions.md)
```

aSemy commented:
Ahhhhh is that why it's been so tricky to create a unit test in the project for complete JDK 17 support? Because the test classes are allowed to create instances of classes within the same package?

aSemy commented on doc/md/jdk16-access-exceptions.md:
````diff
@@ -0,0 +1,62 @@
+   ```
+
+### Problem cause
+JDK 16 enforced strong encapsulation of standard modules. In practice this means you (and your libraries) can no more get away with call like `privateMember.setAccessible(true)` to any of JDK classes.
````
```suggestion
JDK 16 enforces strong encapsulation of standard modules. In practice this means libraries are forbidden from overriding accessibility modifiers in JDK classes (for example,  `privateMember.setAccessible(true)`).
```

sergey-volkov-lm replied:
> In practice this means libraries are forbidden

AFAIK not only libraries. Do you want to make accent on mockk use-case?

aSemy replied:
Good point, I've edited it. 

I think the project name should be styled as: MockK

aSemy commented on doc/md/jdk16-access-exceptions.md:
```diff
@@ -0,0 +1,62 @@
+https://blogs.oracle.com/javamagazine/post/a-peek-into-java-17-continuing-the-drive-to-encapsulate-the-java-runtime-internals
+
+### Solution / workaround
+Add JVM argument `--add-opens java.base/java.time=ALL-UNNAMED` _for each package_ you need  
```
```suggestion
Add JVM argument `--add-opens java.base/java.time=ALL-UNNAMED` _for each package_ you need to mock.
```

sergey-volkov-lm replied:
Agree.
And to be more general here, we also can replace `java.time` with `package.name`. And there are specific examples later in the file

aSemy replied:
Looking at the [docs here](https://docs.oracle.com/en/java/javase/16/migrate/migrating-jdk-8-later-jdk-releases.html#GUID-2F61F3A9-0979-46A4-8B49-325BA0EE8B66), would the flag `--illegal-access=warn` work better?

sergey-volkov-lm replied:
They say `--illegal-access` was removed in JDK 17

[Article](https://blogs.oracle.com/javamagazine/post/a-peek-into-java-17-continuing-the-drive-to-encapsulate-the-java-runtime-internals), section "What’s changing in Java 17?"

Because of this I did not mention this flag. I think that JDK 17 is going to be much more popular, and current users of JDK 16 should not rely on `--illegal-access` option, in order not to face access problems again in future.

aSemy replied:
Ah clear, thanks for explaining.

aSemy reviewed (approved):
Nice work :)

I have some minor formatting and grammar tweaks.

I'm approving, but I'm not certain how the website works - but that can be fixed later if the file isn't accessible.

sergey-volkov-lm commented on doc/md/jdk16-access-exceptions.md:
````diff
@@ -0,0 +1,62 @@
+Add JVM argument `--add-opens java.base/java.time=ALL-UNNAMED` _for each package_ you need  
+Description: https://docs.oracle.com/en/java/javase/16/migrate/migrating-jdk-8-later-jdk-releases.html
+
+For Gradle users:
+```groovy
+tasks.test {
+    jvmArgs("--add-opens", "java.base/java.time=ALL-UNNAMED")
+}
+```
````
I am not very familiar with gradle, and copied this from one of the issues. May need help from experienced gradle user to verify that this snippet looks ok

aSemy replied:
Yeah it looks okay. It will only change the 'test' class though. In .kts, this is the best way to change all test tasks:

```kts
// build.gradle.kts

tasks.withType<Test>().configureEach {
  jvmArgs(...)
}
```

Raibaz commented on doc/md/jdk16-access-exceptions.md:
```diff
@@ -0,0 +1,62 @@
+# JDK 16+ access exceptions
+
+### Problem symptoms 
+On JDK 16 and above you may encounter `InaccessibleObjectException` or `IllegalAccessException` in following known cases:
```
"in the following known cases"

Raibaz reviewed (approved):
This is brilliant, thanks a lot!

Added just a minor grammar nit and I agree with @aSemy's comments and clarifications, but then we can merge it.

I'm not 101% familiar with the website generation so I don't know if linking to a .md file will actually render it or just show the markdown, but I can fix it once we've merged this PR.

Thanks again!

aSemy commented:
Is it possible for MockK to detect, during runtime, if the correct flags are set? Then MockK could print a warning, directing users to these docs.

sergey-volkov-lm commented:
Sorry for delay. I have fixed all what was clear to me. And added additional details because of misunderstanding here https://github.com/mockk/mockk/issues/981

Raibaz commented:
Thanks!
